display.py

```diff
@@ -21,271 +24,309 @@ def font(family='', bold=False, italic=False):
+                        fn(edge_ctx)
+            for point in grid.points:
+                fn, matrix = self._setup_point(point)
+                if fn:
+                    with transform_drawing_context(ctx, matrix):
+                        point_ctx = PointContext(self, ctx, point, model, extra)
+                        fn(point_ctx)
+
+    def _get_surface(self, grids, scale):
+        def adjust(extents, x, y):
+            l, t, r, b = extents
+            return l+x, t+y, r+x, b+y
+
+        lefts, tops, rights, bottoms = tuple(zip(*[adjust(self._get_extents(grid), x, y) for grid, x, y, *_ in grids]))
+        left = min(lefts)
+        top = min(tops)
+        right = max(rights)
+        bottom = max(bottoms)
+
+        # add padding on every side
+        width = (right - left + self.padding * 2) * scale
+        height = (bottom - top + self.padding * 2) * scale
+        surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, int(width), int(height))
+        ctx = cairo.Context(surface)
+        ctx.set_source_rgba(1, 1, 1, 1)
+        ctx.rectangle(0, 0, width, height)
+        ctx.fill()
 
-def canvas_size(w, h, scale):
-    return (w+1) * scale, (h+1) * scale
+        # translate context so user origin and grid origin coincide
+        ctx.scale(scale, scale)
+        ctx.translate(-left + self.padding, -top + self.padding)
+        return surface, ctx
 
-class PointContext(object):
-    def __init__(self, surface, point, model, scale):
-        self.surface = surface
-        self.point = point
-        self.model = model
-        self.scale = scale
-        self.ctx = cairo.Context(surface)
+    def _get_extents(self, grid):
+        raise NotImplementedError
 
-    @property
-    def val(self):
-        return str(self.model[self.point.var])
+    def _setup_cell(self, cell):
+        raise NotImplementedError
 
-    @property
-    def x0(self):
-        return transform_x(self.gx, self.scale)
+    def _setup_edge(self, edge):
+        raise NotImplementedError
 
-    @property
-    def y0(self):
-        return transform_y(self.gy, self.scale)
+    def _setup_point(self, point):
+        raise NotImplementedError
 
-    @property
-    def gx(self):
-        return self.point.x
+    def cell_corners(self):
+        raise NotImplementedError
 
-    @property
-    def gy(self):
-        return self.point.y
+    # override this if the cells in this grid type are not unit circles; it should be the radius of something
+    # that will fit entirely inside a cell
+    CELL_RADIUS = 1
+
+    def convert_coords(self, coords, default):
+        raise NotImplementedError
 
-    def draw_square(self, size=1, color=(0, 0, 0, 1)):
-        r, g, b, a = color
-        self.ctx.set_source_rgba(b, g, r, a)
-        self.ctx.rectangle(self.x0 - size/2, self.y0 - size/2,
+SQRT2 = math.sqrt(2)
+# users would much rather specify stroke widths and text sizes in device coordinates, instead of user coordinates.
+def convert_stroke_width(ctx, width):
```

Comment:
This is a perfect example of why I am against using Cairo transforms for all of our scaling needs. It works well if we want to uniformly scale everything, but this pre-un-scaling so that Cairo scaling results in an unscaled value dance stinks. Introducing floating-point math to compensate for what Cairo is doing under the hood, versus just doing simple integer multiplication on our end for vertex positioning? The latter is way clearer and less error-prone.

Reply:
God I went back and forth on this SO MANY TIMES. Would you be happier specifying line widths and text sizes in terms of `scale`, just like everything else is? I could certainly get behind that. Because I agree, I didn't like this solution either but it was the best I could come up with. This also should have been `_convert_stroke_width()`; I'm not sure what happened to its underscore.

Reply:
Like I said in the top-level comment, I think you made the right call with respect to the user interface on this one—line widths and text sizes should be specified in device scale. It's the implementation that I object to.